contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,268 @@
+
+contract RockPaperScissors is Ownable, Pausable {
+
+    using SafeMath for uint;
+
+    mapping(address => uint) public balances;
+    mapping(bytes32 => Game) public games;
+
+    address private constant NULL_ADDRESS = address(0);
+    string private constant ONLY_CREATOR_CALL_MSG = "Only creator can call";
+    string private constant GAME_NOT_FOUND_MSG = "Game not found";
+    string private constant INVALID_MOVE_MSG = "Invalid move";
+
+    uint256 constant FORFEIT_WINDOW = 12 hours;
+    uint256 constant FEE_PERCENTAGE = 10;
+
+    struct Game {
+        address creator;
+        uint256 stake;
+        bytes32 moveHash;
+        address opponent;
+        uint8 opponentMove;
+        uint256 expiryDate;
+    }
+
+    event Deposit(
+        address player,
+        uint256 amount
+    );
+
+    event GameCreated(
+        address indexed creator,
+        bytes32 indexed gameToken,
+        uint256 stake
+    );
+
+    event OpponentPlays(
+        address indexed opponent,
+        bytes32 indexed token,
+        uint8 move
+    );
+
+    event Draw(
+        bytes32 indexed token
+    );
+
+    event WinnerRewarded(
+        bytes32 indexed gameToken,
+        address indexed winner,
+        address indexed loser,
+        uint256 winnings,
+        uint8 creatorMove,
+        uint8 opponentMove
+    );
+
+    event WithDraw(
+        address indexed withdrawer,
+        uint amount
+    );
+
+    event  CreatorEndsGame(
+        address indexed player,
+        bytes32 indexed gameToken,
+        uint256 refund
+    );
+
+    event  ForfeitPaid(
+        address indexed player,
+        bytes32 indexed gameToken,
+        uint256 forfeit
+    );
+
+    function deposit() external payable whenNotPaused {
+
+        balances[msg.sender] = balances[msg.sender].add(msg.value);
+
+        emit Deposit(msg.sender, msg.value);
+    }
+
+    function create(bytes32 playerMoveHash, uint stake) external payable whenNotPaused  returns(bytes32){
+
+        require(playerMoveHash != bytes32(0), "Player move cannot be empty");
+        require(stake > 0, "Stake cannot be zero");
+
+        bytes32 gameToken = keccak256(abi.encodePacked(msg.sender, block.timestamp));
+
+        Game storage game = games[gameToken];
+        require(game.creator == NULL_ADDRESS, "Token already in use");
+
+        game.creator = msg.sender;
+        game.stake = stake;
+        game.moveHash = playerMoveHash;
+        fundMove(stake);
+        emit GameCreated(msg.sender, gameToken, stake);
+
+        return gameToken;
+    }
+
+    /*
+    * The opponent makes their move and waits for the creator to resolve the game by revealing their move
+    */
+    function opponentPlay(bytes32 gameToken, uint8 playerMove) external payable whenNotPaused {
+
+        Game storage game = games[gameToken];
+        uint256 stake = game.stake;
+        require(stake != 0, GAME_NOT_FOUND_MSG);
+        require(game.opponent == NULL_ADDRESS, "Opponent has already moved");
```

Comment:
Do you see how player1 wins 95% of the time?

Reply:
Oh. Because I forgot to add this line further down?

`game.opponent = msg.sender`

Reply:
Not for this reason.

Reply:
I have thought long and hard but can't see it. Perhaps it will become more apparent when I write tests.

Reply:
It will not appear in tests unless you specifically test for it.
What does `require(game.opponent == NULL_ADDRESS` mean when combined with the pool of pending transactions?

Reply:
Sorry. I am stuck. All I can understand is you are suggesting that there is some way for the creator to overwrite the opponent's move by resetting the opponent address back to the zero address but I cannot see how that can be done.

Reply:
Is the vulnerability potentially a transaction-ordering attack?

Reply:
I think what I need to do is to establish the opponent's address in storage before the move reveal step. The move reveal step would then be checked against the stored opponent's address.

Reply:
Reveal for me is when player 1 reveals the salt and the move.
Here, for player 2, the move is there from the start. So while the transaction is waiting in the transaction pool to be mined, player 1 can see it, and see what player 2 is playing.
At this moment, player 1 knows whether they will win or lose. If they know they lose, player 1 can quickly send a transaction with a higher gas price from their other address with whatever move. After, it is a resolution between player 1 and player 1. Player 2's transaction will revert, preventing them to play. 
So the 95% chance is a matter of gas price difference and mining speed.

Reply:
Thank you. I understand now. I will add and extra step to get the opponent to register to play and store their address before they reveal . They then reveal but it must be matched against the stored address. This will prevent the creator front-running.

Reply:
Having a given my last proposal more thought, I realised that it doesn't remove the problem and that I have to get the creator to reveal before the opponent reveals.

Reply:
I think if player2 joins first, and then reveals the move, that's ok because player1 cannot front run player2 since player2 address is stored.